src/shogun/machine/GLM.h

```diff
@@ -0,0 +1,268 @@
+#include <shogun/machine/Machine.h>
+#include <shogun/machine/LinearMachine.h>
+#include <shogun/machine/IterativeMachine.h>
+#include <shogun/mathematics/RandomMixin.h>
+#include <shogun/distributions/Distribution.h>
+#include <shogun/optimization/GradientDescendUpdater.h>
+#include <shogun/optimization/ConstLearningRate.h>
+#include <shogun/optimization/ElasticNetPenalty.h>
+#include <shogun/mathematics/linalg/LinalgNamespace.h>
+#include <shogun/optimization/FirstOrderCostFunction.h>
+
+namespace shogun
+{
+    enum GLM_DISTRIBUTION
+	{
+        //TODO GAUSSIAN,
+        //TODO BINOMIAL,
+        //TODO GAMMA,
+        //TODO SOFTPLUS,
+        //TODO PROBIT,
+        POISSON
+	};
+
+class DotFeatures;
+class Features;
+class RegressionLabels;
+class GLMCostFunction;
+
+/** @brief Class for estimating regularized generalized linear models (GLM).
+ *   The regularized GLM minimizes the penalized negative log likelihood:
+ *
+ *  This uses Elastic-net penalty which defaults to the ridge penalty when
+ *  alpha = 0 and defaults to the lasso penalty when alpha = 1.
+ *
+ * */
+class GLM : public RandomMixin<IterativeMachine<LinearMachine>>
+{
+ 	public:
+		friend class GLMCostFunction;
+		
+		/** default constructor */
+		GLM();
+
+ 		GLM(GLM_DISTRIBUTION distribution, float64_t alpha, float64_t lambda, float64_t learning_rate, int32_t max_iterations, float64_t tolerance, float64_t eta);
+
+ 		/** destructor */
+ 		~GLM() override {}
+
+		/** apply linear machine to data
+		 * for regression problem
+		 *
+		 * @param data (test)data to be classified
+		 * @return classified labels
+		 */
+		std::shared_ptr<RegressionLabels> apply_regression(std::shared_ptr<Features> data=NULL) override;
+
+ 		/** Returns the name of the SGSerializable instance.  It MUST BE
+ 		 *  the CLASS NAME without the prefixed `C'.
+ 		 *
+ 		 * @return name of the SGSerializable
+ 		 */
+ 		const char* get_name() const override { return "GLM"; }
+
+		void set_tau(SGMatrix<float64_t> tau);
+
+		SGMatrix<float64_t> get_tau();
+
+	protected:
+
+		void init_model(const std::shared_ptr<Features>& data);
+		
+		void iteration() override;
+
+	protected:
+
+		/** Distribution type */
+		GLM_DISTRIBUTION distribution = POISSON;
+
+		/** a threshold parameter that linearizes the exp() function above eta. */
+		float64_t m_eta = 2.0;
+
+		/** regularization parameter :math:`\\lambda` of penalty term. */
+		float64_t m_lambda = 0.1;
+
+		/** the weighting between L1 penalty and L2 penalty term of the loss function. */
+		float64_t m_alpha = 0.5;
+
+		/** convergence threshold or stopping criteria. Optimization loop will stop when relative change in parameter norm is below the threshold. */
+		float64_t m_tolerance = 1e-6;
+
+		bool m_compute_bias = true;
+
+		std::shared_ptr<GradientDescendUpdater> m_gradient_updater;
+
+		std::shared_ptr<ConstLearningRate> m_learning_rate;
+
+		std::shared_ptr<ElasticNetPenalty> m_penalty;
+
+		std::shared_ptr<GLMCostFunction> m_cost_function;
```

Comment:
Anyhow, I guess the segmentation fault could be caused by these guys here. They are not initialized anywhere in the code. Therefore, if you try to access one of their methods, the program will segfault.

Reply:
I tried initializing them in the `GLM()` default constructor, but as it seems program control isn't entering the `GLM()` constructor(As much as I could deduce from logs using cout). And I still get the segfault.

Reply:
initialise them down here, and all should be good

Reply:
the shared_ptr have to be initialise in the constructor, as they allocate memory on the heap (compiler cannot allocate this memory as part of object initialisation on the stack afaik)

Reply:
Yes thanks, I have done this now, and it works.